### DISCUSSION TITLE
Neko on LUMI-G

### DISCUSSION THREAD

**Discussion by njansson** [ExtremeFLOW/neko#665] [Show and tell]
### To compile Neko on LUMI-G
_(For a general guide on how to build Neko for AMD GPUs please see: https://github.com/ExtremeFLOW/neko/discussions/585)_
* Make sure have to have the following modules loaded:
   - The Cray software stack (`ml CrayEnv`)
   - Cray Accelerator Target (`ml craype-accel-amd-gfx90a`)
   - ROCm Toolkit (`ml rocm`) 
* Two different compiler combinations can be used:
 1. Configure Neko to use Cray Fortran + HIPCC and device aware MPI :
```shell
./configure  FC=ftn CC=cc --with-hip HIP_HIPCC_FLAGS="-O3 --offload-arch=gfx90a" --enable-device-mpi
```
 2. Configure Neko to use Cray Fortran + CRAY C++ and device aware MPI : 
```shell
./configure  FC=ftn CC=cc --with-hip HIPCC=CC HIP_HIPCC_FLAGS="-O3 -x hip --offload-arch=gfx90a" --enable-device-mpi
```
* Build using `make && make install`
* Make sure to set `MPICH_GPU_SUPPORT_ENABLED=1` in the shell/jobscript before running `neko` (_**otherwise device aware MPI will cause a segmentation fault**_), if everything went well backend type should be reported as HIP (but currently not the device type...).
```
    _  __  ____  __ __  ____ 
   / |/ / / __/ / //_/ / __ \
  /    / / _/  / ,<   / /_/ /
 /_/|_/ /___/ /_/|_|  \____/ 
 
 (version: 0.4.99)
 (build: 2022-10-04 on x86_64-pc-linux-gnu using cray)
 
  
 -------Job Information-------- 
 Start time: 23:01 / 2022-10-04
 Running on: 32 MPI ranks
 Bcknd type: Accelerator (HIP)
 Dev. name :
 Real type : double precision
```

**Comment by ronithstanly** [comment]
After the latest Lumi summer updates the following works:

`ml CrayEnv cpe/23.09 craype-accel-amd-gfx90a rocm`

`./configure MPIFC=ftn MPICC=cc MPICXX=CC CFLAGS="-O3" FCFLAGS="-O3" --prefix=/path/to/neko_install HIPCC=hipcc FC=ftn CC=cc --with-hip=${ROCM_PATH} HIP_HIPCC_FLAGS="-O3 --offload-arch=gfx90a" --enable-device-mpi`

`make && make install`

**Reply by MartinKarp** [reply]
Slight update: (`-O2` instead of `-O3` for the FCFLAGS)

` ./configure MPIFC=ftn MPICC=cc MPICXX=CC CFLAGS="-O3" FCFLAGS="-O2" --prefix=/path/to/neko_install HIPCC=hipcc FC=ftn CC=cc --with-hip=${ROCM_PATH} HIP_HIPCC_FLAGS="-O3 --offload-arch=gfx90a" --enable-device-mpi`

**Comment by shiyu-du-dsy** [comment]
After the maintenance of LUMI in January 2026, the following works
```
ml CrayEnv cce/19.0.0 craype-accel-amd-gfx90a rocm/6.3.4
export LD_LIBRARY_PATH=$CRAY_LD_LIBRARY_PATH:$LD_LIBRARY_PATH

./regen.sh

./configure MPIFC=ftn MPICC=cc MPICXX=CC CFLAGS="-O3" FCFLAGS="-O0" --prefix=$PWD HIPCC=hipcc FC=ftn CC=cc --with-hip=${ROCM_PATH} HIP_HIPCC_FLAGS="-O3 --offload-arch=gfx90a" --enable-device-mpi

make -j && make install
```
The optimisation level of Fortran compiler needs to be zero otherwise it reports segmentation fault when setting up the connectivity. cce/19.0.0 is not necessary to load but consider they are marching towards cce/20.0.0, it's better to nail the setup here.

**Reply by njansson** [reply]
19 and 20 has some issues, if possible swap back to 18

**Reply by shiyu-du-dsy** [reply]
Unfortunately it seems that they don't have cce 18.

**Reply by vbaconnet** [reply]
~~Works fine for me with `-O2`. Just loaded `module load craype-accel-amd-gfx90a rocm`, and the same configure as you.~~

**Reply by vbaconnet** [reply]
Sorry, please disregard my message above. I can replicate the segfault as you describe it.

**Reply by njansson** [reply]
I'm working on resolving this, but progress is slow :(

**Reply by shiyu-du-dsy** [reply]
I am trying with a case with around 34k elements with polynomial order 7 on 1 node with 8 devices, and it works. However, when I shift the same case to around 40k elements with the same polynomial order, it reports OOM when setting up GS. Not sure whether it's related to this known issue or not.

**Reply by njansson** [reply]
Have you checked affinity mappings etc, such that everything doesn't end up on the same GCD? Furthermore, good to check the load balance of the mesh, such that there isn't excessive amount of off -ranks connections

**Reply by shiyu-du-dsy** [reply]
no, how should I check them? Maybe LUMI website?

**Reply by njansson** [reply]
Yes, there should be affinity setting examples on LUMI's support page

**Reply by shiyu-du-dsy** [reply]
okay thank you!

**Reply by mszpindler** [reply]
After the update Slurm accepts following binding options:
```
#SBATCH --ntasks=8
#SBATCH --ntasks-per-node=8
#SBATCH --exclusive
#SBATCH --partition=standard-g
#SBATCH --gpus-per-node=8

srun --cpus-per-task=7 --gpu-bind=map:4,5,2,3,6,7,0,1 --gres-flags=allow-task-sharing ./neko your.case
```

**Reply by njansson** [reply]
> I'm working on resolving this, but progress is slow :(

There's a patch now in the `develop` (see #2434) that "resolves" the issue, apply it using `git apply patches/cce_stack.patch`

In short, CCE does some really aggressive optimisations that destroys stack growth, for a more detailed description see #2209.

**Comment by loretet** [comment]
UPDATE

As of April 21 2026 this configuration seems to work.

Modules:
```
ml CrayEnv cce/19.0.0 craype-accel-amd-gfx90a rocm/6.3.4 cray-python
export JSON_INSTALL=/users/<USERNAME>/json-fortran/b
export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:${JSON_INSTALL}/lib64/ #On some systems lib should be replaced with lib64
export PKG_CONFIG_PATH=${PKG_CONFIG_PATH}:${JSON_INSTALL}/lib64/pkgconfig
export LD_LIBRARY_PATH=$CRAY_LD_LIBRARY_PATH:$LD_LIBRARY_PATH
```

Configuration:
```
./configure --prefix=$PWD MPIFC=ftn MPICC=cc MPICXX=CC CFLAGS="-O3" FCFLAGS="-O0" HIPCC=hipcc FC=ftn CC=cc --with-hip=${ROCM_PATH} HIP_HIPCC_FLAGS="-O3 --offload-arch=gfx90a" --enable-device-mpi --with-adios2
```

Run script (big run):
```
#!/bin/bash -l
#SBATCH --job-name=Neko
#SBATCH --partition=dev-g
#SBATCH --nodes=128
#SBATCH --ntasks-per-node=8
#SBATCH --gpus-per-node=8
#SBATCH --time=0-00:05:00               
#SBATCH --mail-type=all                 
#SBATCH -c 6
#SBATCH --account=project_****

cat << EOF > select_gpu
#!/bin/bash

export ROCR_VISIBLE_DEVICES=\$SLURM_LOCALID
exec \$*
EOF

chmod +x ./select_gpu

CPU_BIND="mask_cpu:7e000000000000,7e00000000000000"
CPU_BIND="${CPU_BIND},7e0000,7e000000"
CPU_BIND="${CPU_BIND},7e,7e00"
CPU_BIND="${CPU_BIND},7e00000000,7e0000000000"

export OMP_NUM_THREADS=2
export MPICH_GPU_SUPPORT_ENABLED=1
export NEKO_GS_STRTGY=3

d="$(date +%F_%H-%M-%S)"

srun -u --cpu-bind=${CPU_BIND},verbose ./select_gpu ./neko CASENAME.case >> log.run{d} 2>&1
rm -rf ./select_gpu
```

**Reply by loretet** [reply]
This is an improved version of the run script which adds some additional checks and switches the CPU bind based on the case size:

```
#!/bin/bash -l
#SBATCH --job-name=Neko
#SBATCH --partition=dev-g
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=8
#SBATCH --gpus-per-node=8   
#SBATCH --time=0-00:05:00              
#SBATCH --mail-type=all                
#SBATCH -c 6                 
#SBATCH --account=project_****

ml CrayEnv cce/19.0.0 craype-accel-amd-gfx90a rocm/6.3.4 cray-python
export USERN="<USERNAME>"
export OMP_NUM_THREADS=2
export MPICH_GPU_SUPPORT_ENABLED=1
export NEKO_GS_STRTGY=3
export JSON_INSTALL=/users/$USERN/json-fortran/b
export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:${JSON_INSTALL}/lib64/
export PKG_CONFIG_PATH=${PKG_CONFIG_PATH}:${JSON_INSTALL}/lib64/pkgconfig
export LD_LIBRARY_PATH=$CRAY_LD_LIBRARY_PATH:$LD_LIBRARY_PATH

if [ "$SLURM_NTASKS_PER_NODE" -ne "$SLURM_GPUS_PER_NODE" ]; then
    echo "Error: Rank-to-GPU mismatch. Tasks: $SLURM_NTASKS_PER_NODE, GPUs: $SLURM_GPUS_PER_NODE"
    echo "--ntasks-per-node must be == to --gpus-per-node"   
    exit 1 
fi

if [ "$OMP_NUM_THREADS" -gt "$SLURM_CPUS_PER_TASK" ]; then
    echo "Error: OMP_NUM_THREADS ($OMP_NUM_THREADS) exceeds CPUs per task ($SLURM_CPUS_PER_TASK)."
    echo "#SBATCH -c must be >= OMP_NUM_THREADS"
    exit 1
fi

if [ "$SLURM_NNODES" -le 2 ]; then
    BIND_SETTING="cores"
    echo "Small case detected. Using automated core binding."
else
    BIND_SETTING="mask_cpu:7e000000000000,7e00000000000000,7e0000,7e000000,7e,7e00,7e00000000,7e0000000000"
    echo "Large case detected. Applying optimized hex masks."
fi

cat << EOF > select_gpu
#!/bin/bash

export ROCR_VISIBLE_DEVICES=\$SLURM_LOCALID
exec \$*
EOF

chmod +x ./select_gpu

d="$(date +%F_%H-%M-%S)"
srun -u --cpu-bind=${BIND_SETTING},verbose ./select_gpu ./neko abl_test.case >> log.run_${d} 2>&1
rm -rf ./select_gpu
```

The `exit 1`'s are there because, at least personally, I'd rather my simulation crash right away than being super slow and consume all my precious resources

**Comment by mszpindler** [comment]
I am afraid there is another problem with CCE 21.0.0 causing pressure solver to diverge if compiler opt level is not 0. This is different from #2434.

**Reply by njansson** [reply]
Hmm, there can of course be some more shenanigans from CCE...with all cce related patches applied and CCE 21 (CPE 26.03) we things looked ok. However, that was for a checkout of develop in mid/late June.

**Reply by njansson** [reply]
Ps.

This recent PR https://github.com/ExtremeFLOW/neko/pull/2614 also fixed some optimisation atrocities with CCE
